util/yang.go

```diff
@@ -361,6 +362,37 @@ func EnumeratedUnionTypes(types []*yang.YangType) []*yang.YangType {
 	return eTypes
 }
 
+// EnumeratedUnionSubsumingTypes returns the subsuming types of the enumerated
+// types of the input union type in tree order of enumerations. This matches
+// the order returned by EnumeratedUnionTypes. The input subsumingType is the
+// union type containing the input union subtypes.
+// The "subsuming type" of each `YangType` means the closest/innermost type
+// that's not a union subtype:
+// case 1: for any type not within a union, the subsuming type is itself.
+// case 2: for any type within a union, the subsuming type is the first
+// containing union type that's a typedef, or highest union type definition,
+// whichever one is closest to the type.
+func EnumeratedUnionSubsumingTypes(subsumingType *yang.YangType) []*yang.YangType {
+	return enumeratedUnionSubsumingTypes(subsumingType.Type, subsumingType)
+}
+
+func enumeratedUnionSubsumingTypes(types []*yang.YangType, subsumingType *yang.YangType) []*yang.YangType {
```

Comment:
I wonder if `[]*yang.YangType` is the best return type here -- it seems like we actually just want a tree that gives us the hierarchy of types that exist for a particular leaf, right? Should we define some form of type tree that we return to the user - such that this data structure is a bit clearer?

In the tests, having repeated `parent-type` made it a little difficult to understand what was being represented there -- we've unrolled a tree, but we've denormalised the output based on traversal approach.

It may be that I'm not quite getting why we need to have this structure -- so I'll come back to this comment after reviewing the functionality in the subsequent CL.

Reply:
I feel like I made the concept of "subsuming type" too complex in this change. I've tried to make its meaning simpler to understand. The reason it's written this way is to make the code easier to write, as the users of this "subsuming type" finder function are the very same places where `enumeratedUnionType` is used.

I've changed the name of the function, it may be too long now, but I mulled over it and couldn't come up with a shorter one that still encapsulates the meaning.

Reply:
As above